### DISCUSSION TITLE
How can I add git info to the footer?

### DISCUSSION THREAD

**Discussion by reece** [squidfunk/mkdocs-material#6402] [Q&A] [answered]
I would like to add the git tag to the footer of rendered pages. Here's what I've tried:

- Browsed all of the mkdocs plugins I could find that dealt with git info. They all seem to make git info available to individual markdown pages, but not to jinja templates.
- I've also tried the mkdocs-macros plugin. `git.tag` is exactly what I want. I created a overrides directory with the footer.html partial and tried adding `{{git.tag}}` but I get  `jinja2.exceptions.UndefinedError: 'git' is undefined`.
- I then read https://github.com/fralau/mkdocs_macros_plugin/issues/83#issuecomment-844263415. It's unclear if that technique works at all, but in any case would require adding the tag to the frontmatter of every page.
- I've considered writing a plugin, but I'm surprised that injecting git info into a footer is difficult, so I'm asking here first.

So, any ideas how I can get git.tag (or similar) into the footer (presumably via jinja templates)?

For clarity, I want something like this:
![image](https://github.com/squidfunk/mkdocs-material/assets/109453/b1899e1d-5f7c-4696-b8e0-6e89d8a49472)

(I fudged this with a static `copyright` config in mkdocs.yml)

**Comment by alexvoss** [comment, marked as answer]
I have just had the same problem and found the answer in the blog plugin, specifically [in this line, which adds data to the page context](https://github.com/squidfunk/mkdocs-material/blob/fd90fa6518f06debc1312b167ae3e94386937dff/src/plugins/blog/plugin.py#L354). No need to write a full plugin, you can write a much simpler [hook](https://www.mkdocs.org/user-guide/configuration/#hooks) that has an `on_page_context` method and adds the data you want.

**Reply by reece** [reply]
Specifically, here's how:

In mkdocs.yml:
```
hooks:
  - hooks.py

theme:
  name: material
  custom_dir: overrides

```

hooks.py:
```
import logging

import git
import pytz

import mkdocs.plugins

_logger = logging.getLogger()

@mkdocs.plugins.event_priority(-100)
def on_env(env, config, files):
    r = git.Repo(".")
    tag = r.git.describe(tags=True)
    hc = r.head.commit
    commit_datetime_utc = r.head.commit.committed_datetime.astimezone(pytz.timezone("UTC"))
    env.globals["git"] = {
        # intended to be consistent with mkdocs-macros-plugin git info:
        "author": hc.author.author,
        "author_email": hc.author.email,
        "commit": hc.hexsha,
        "committer": hc.committer.name,
        "committer_email": hc.committer.email,
        "date": hc.committed_date,
        "message": hc.message,
        "short_commit": hc.hexsha[:7],
        "tag": tag,

        # bonus:
        "datetime_utc": commit_datetime_utc,
        }
    _logger.info(f"Entering on_env hook; added {tag}")
    return env
```

overrides/partials/version.html:
```
{% if "-" in git.tag %}
{% set repo_url = config.repo_url + "/commit/" + git.commit %}
{% else %}
{% set repo_url = config.repo_url + "/releases/tag/" + git.tag %}
{% endif %}

<a href="{{ repo_url }}">
  <div>
    {{ git.tag }}
    <br/>{{ git.datetime_utc.date() }}
  </div>
</a>
```

Excerpt of overrides/partials/header.html:
copied from venv/lib/python3.11/site-packages/material/templates/partials/header.html
```
⋮
      {% include "partials/search.html" %}
    {% endif %}
    {# Disabled by Reece; display custom version info
    {% if config.repo_url %}
      <div class="md-header__source">
        {% include "partials/source.html" %}
      </div>
    {% endif %} #}
      <div class="md-header__source">
        {% include "partials/version.html" %}
      </div>
  </nav>
⋮
```

Which looks like this:
![image](https://github.com/squidfunk/mkdocs-material/assets/109453/6eb9f61a-377e-445a-9398-32b1b7d089ba)


(From https://hgvs-nomenclature.org/)
